middleman_protocol/middleman_protocol/message.py

```diff
@@ -0,0 +1,302 @@
+        if not isinstance(payload, BaseGolemMessage):
+            raise MiddlemanProtocolError(
+                f'Trying to create GolemMessageMiddlemanMessage but passed payload type is {type(payload)} '
+                f'instead of Golem Message.'
+            )
+
+
+@register
+class ErrorMiddlemanMessage(AbstractMiddlemanMessage):
+    """ Subclass of AbstractMiddlemanMessage for handling error code with error message payload type. """
+
+    payload_type = PayloadType.ERROR
+
+    @classmethod
+    def _deserialize_payload(cls, payload: bytes) -> tuple:
+        error_payload_format = cls.get_error_payload_format()
+        error_payload = error_payload_format.parse(payload)
+        return (error_payload.error_code, error_payload.error_message)
+
+    def _serialize_payload(self, payload: tuple) -> bytes:
+        error_payload_format = self.get_error_payload_format()
+        error_payload = Container(
+            error_code=payload[0],
+            error_message=payload[1],
+        )
+        raw_error_payload = error_payload_format.build(error_payload)
+        return raw_error_payload
+
+    def _validate_payload(self, payload: tuple) -> None:
+        if not isinstance(payload, tuple):
+            raise MiddlemanProtocolError(
+                f'Trying to create ErrorMiddlemanMessage but passed type payload is '
+                f'{type(payload)} instead of tuple. It must be pair of error code and error message.'
+            )
+        if len(payload) != 2:
+            raise MiddlemanProtocolError(
+                f'Trying to create ErrorMiddlemanMessage but passed payload tuple has length '
+                f'{len(payload)} instead of 2. It must be pair of error code and error message.'
+            )
+        if not isinstance(payload[0], int):
+            raise MiddlemanProtocolError(
+                f'First element of payload tuple passed to ErrorMiddlemanMessage must be error code integer '
+                f'instead of {type(payload[0])}.'
+            )
+        if not isinstance(payload[1], str):
+            raise MiddlemanProtocolError(
+                f'Second element of payload tuple passed to ErrorMiddlemanMessage must be error message string '
+                f'instead of {type(payload[1])}.'
+            )
+
+    @classmethod
+    def get_error_payload_format(cls) -> Struct:
+        return Struct(
+            error_code=Int8ub,
+            error_message=PascalString(VarInt, 'utf8'),
+        )
+
+
+@register
+class AuthenticationChallengeMiddlemanMessage(AbstractMiddlemanMessage):
+    """ Subclass of AbstractMiddlemanMessage for handling authentication challenge payload type. """
+
+    payload_type = PayloadType.AUTHENTICATION_CHALLENGE
+
+    @classmethod
+    def _deserialize_payload(cls, payload: bytes) -> bytes:
+        return payload
+
+    def _serialize_payload(self, payload: bytes) -> bytes:
+        return payload
+
+    def _validate_payload(self, payload: tuple) -> None:
+        pass
```

Comment:
If it's not finished, please raise `NotImplementedError` or do `assert False`.

Reply:
It works differently. The subclasses of `AbstractMiddlemanMessage` require this method to be implemented. It is always called when subclass is instantiated. But currently there are no checks to be performed, it will accept any `bytes` (I guess it will be defined and updated in next issues).

Btw. there should be `bytes` in signature not `tuple`. :+1: done

Reply:
OK.